mesonbuild/mcompile.py

```diff
@@ -0,0 +1,136 @@
+        '-l', '--load-average',
+        action='store',
+        default=0,
+        type=int,
+        help='The system load average to try to maintain (if supported)'
+    )
+    parser.add_argument(
+        '--debug',
+        action='store_true',
+        help='Print debug information for the meson compile command.'
+    )
+    parser.add_argument(
+        '--clean',
+        action='store_true',
+        help='Clean the build directory.'
+    )
+    parser.add_argument(
+        'builddir',
+        type=pathlib.Path,
+        nargs='?',
+        default='.',
+        help='The directory containing build files to be built.'
+    )
+
+
+def _run(options: 'argparse.Namespace') -> int:
+    if options.debug:
+        mlog.log_file = sys.stderr
+        mlog.set_verbose()
+
+    bdir = options.builddir  # type: pathlib.Path
+    if not bdir.exists():
+        mlog.error('Path to builddir {} does not exist!'.format(bdir.as_posix()))
+    if not bdir.is_dir():
+        mlog.error('builddir path should be a directory.')
+
+    cmd = []  # type: T.List[str]
+    runner = None  # type T.Optional[str]
+    if (bdir / 'build.ninja').exists():
+        mlog.debug('Found build.ninja, using ninja.')
+        runner = os.environ.get('NINJA')
+        mlog.debug('Got runner {} from environment variable NINJA'.format(runner))
+        if not runner:
+            if shutil.which('ninja'):
+                runner = 'ninja'
+            elif shutil.which('samu'):
+                runner = 'samu'
+            mlog.debug('found runner {}'.format(runner))
+
+        if runner is None:
+            mlog.error('Cannot find either ninja or samu.')
+            return 1
+
+        cmd = [runner, '-C', bdir.as_posix()]
+
+        # If the value is set to < 1 then don't set anything, which let's
+        # ninja/samu decide what to do.
+        if options.jobs > 0:
+            cmd.extend(['-j', str(options.jobs)])
+        if options.load_average > 0:
+            cmd.extend(['-l', str(options.load_average)])
+        if options.clean:
+            cmd.append('clean')
+
+    # TODO: with python 3.8 this could be `elif sln := bdir.glob('*.sln'):`
+    elif list(bdir.glob('*.sln')):
+        sln = list(bdir.glob('*.sln'))[0]
+        mlog.debug('Found {}, using msbuild.')
+        cmd = ['msbuild', str(sln.resolve())]
+
+        # In msbuild `-m` with no number means "detect cpus", the default is `-m1`
+        if options.jobs > 0:
+            cmd.append('-m{}'.format(options.jobs))
+        else:
+            cmd.append('-m')
+
+        if options.load_average:
+            mlog.warning('Msbuild does not have a load-average switch, ignoring.')
+        if options.clean:
+            cmd.extend(['/t:Clean'])
+
+    # TODO: xcode?
+    else:
+        mlog.error('Could not find any runner or backend for directory {}'.format(bdir.resolve().as_posix()))
+        return 1
+
+    mlog.debug('Running command {}'.format(cmd))
+    p = subprocess.run(cmd)
```

Comment:
Should we use mesonlib.Popen_safe() here?

Reply:
I guess we can, but I don't think it's necessary. We're not passing anything to stdin or capturing the output, we're just letting it be redirected to the terminal.

Reply:
I don't remember what are the extra hacks we have in Popen_safe(), maybe it's not required here.

Reply:
I'll just use Popen_safe, so that the next person to look at this doesn't ask "why didn't they use popen_safe?" and I don't have to say "I don't remember, that was ... ago."